hazelcast/src/main/java/com/hazelcast/topic/impl/TotalOrderedTopicProxy.java

```diff
@@ -47,4 +65,41 @@ public void publish(@Nonnull  E message) {
         InternalCompletableFuture f = invokeOnPartition(operation);
         f.joinInternal();
     }
+
+    @Override
+    public CompletionStage<Void> publishAsync(@Nonnull E message) {
+        checkNotNull(message, NULL_MESSAGE_IS_NOT_ALLOWED);
+        Operation op = new PublishOperation(getName(), toData(message))
+                .setPartitionId(partitionId);
+        return newDelegatingFuture(serializationService, publishInternalAsync(op));
+    }
+
+    @Override
+    public void publishAll(@Nonnull Collection<? extends E> messages) {
+        checkNotNull(messages, NULL_MESSAGE_IS_NOT_ALLOWED);
+        checkNoNullInside(messages, NULL_MESSAGE_IS_NOT_ALLOWED);
+
+        Operation op = new PublishAllOperation(getName(), toDataArray(messages));
+        publishInternalAsync(op).joinInternal();
+    }
+
+    @Override
+    public CompletionStage<Void> publishAllAsync(@Nonnull Collection<? extends E> messages) {
+        checkNotNull(messages, NULL_MESSAGE_IS_NOT_ALLOWED);
+        checkNoNullInside(messages, NULL_MESSAGE_IS_NOT_ALLOWED);
+
+        Operation op = new PublishAllOperation(getName(), toDataArray(messages));
+        return newDelegatingFuture(serializationService, publishInternalAsync(op));
+    }
+
+    private InternalCompletableFuture<Data> publishInternalAsync(Operation operation) {
```

Comment:
Since none of the operations that are passed as arguments to this method actually have a `Data` result (or even have any kind of result), you can change it to `InternalCompletableFuture<Void>`. Also, then you don't need to wrap it in `newDelegatingFuture(serializationService, publishInternalAsync(op))` as the serialization service doesn't have anything to deserialize.

In the end, we then get this:
```java
    @Override
    public CompletionStage<Void> publishAsync(@Nonnull E message) {
        checkNotNull(message, NULL_MESSAGE_IS_NOT_ALLOWED);
        Operation op = new PublishOperation(getName(), toData(message))
                .setPartitionId(partitionId);
        return publishInternalAsync(op);
    }

    @Override
    public void publishAll(@Nonnull Collection<? extends E> messages) {
        checkNotNull(messages, NULL_MESSAGE_IS_NOT_ALLOWED);
        checkNoNullInside(messages, NULL_MESSAGE_IS_NOT_ALLOWED);
        Version clusterVersion = getNodeEngine().getClusterService().getClusterVersion();
        // RU_COMPAT_4_0
        if (!clusterVersion.isGreaterOrEqual(Versions.V4_1)) {
            throw new UnsupportedOperationException("Publish all is not available on cluster version " + clusterVersion
                    + ". Please upgrade the cluster version to " + Versions.V4_1);
        }

        Operation op = new PublishAllOperation(getName(), toDataArray(messages));
        publishInternalAsync(op).joinInternal();
    }

    @Override
    public InternalCompletableFuture<Void> publishAllAsync(@Nonnull Collection<? extends E> messages) {
        checkNotNull(messages, NULL_MESSAGE_IS_NOT_ALLOWED);
        checkNoNullInside(messages, NULL_MESSAGE_IS_NOT_ALLOWED);
        Version clusterVersion = getNodeEngine().getClusterService().getClusterVersion();
        // RU_COMPAT_4_0
        if (!clusterVersion.isGreaterOrEqual(Versions.V4_1)) {
            throw new UnsupportedOperationException("Publish all is not available on cluster version " + clusterVersion
                    + ". Please upgrade the cluster version to " + Versions.V4_1);
        }

        Operation op = new PublishAllOperation(getName(), toDataArray(messages));
        return publishInternalAsync(op);
    }

    private InternalCompletableFuture<Void> publishInternalAsync(Operation operation) {
        topicStats.incrementPublishes();
        try {
            return operationService.invokeOnPartition(OperationService.SERVICE_NAME, operation, partitionId);
        } catch (Throwable t) {
            throw rethrow(t);
        }
    }
```

Reply:
Ok, makes sense. Changed it.